### PR TITLE
add Flink HA document and example

### PR DESCRIPTION
**Pull request by showuon**
Add a `ha-example` folder to demonstrate the Flink HA configuration and usage.

### CODE REVIEW

File: ha-example/README.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+## Flink operator
+
+The Flink operator manages and operates Flink Deployments. It supports high availability through leader election and standby operator instances.
+To enable leader election we need to add the following two mandatory operator configuration parameters.
+
+```
+kubernetes.operator.leader-election.enabled: true
+kubernetes.operator.leader-election.lease-name: flink-operator-lease
+```
+The lease name must be unique in the current lease namespace.
```
Comment (tomncooper) [inline ha-example/README.md:15]: What is the "lease namespace". Is it the ns where the operator is deployed?

File: ha-example/README.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+So, when installing flink operator using helm, we need to add these 2 configs + existing default configs via command like this:
```
Comment (tomncooper) [inline ha-example/README.md:17]:
```suggestion
When installing the Flink Kubernetes operator using `helm`, we need to add these 2 parameters into the existing default configuration string.
This can be done via the command line using the `--set` flag:
```
I realise this is a consequence of how the config passing is set up in the operator, but this is not great UX. If these other options are truely _default_ then they should be in the `helm/flink-kubernetes-operator/conf/flink-conf.yaml`. That way only user supplied defaults could be set via the helm cli/custom values file.

Might be a good improvement to the helm chart upstream.

File: ha-example/README.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+--set replicas=2 \
+--set defaultConfiguration."flink-conf\.yaml"="kubernetes.operator.metrics.reporter.slf4j.factory.class\:\ org.apache.flink.metrics.slf4j.Slf4jReporterFactory
+kubernetes.operator.metrics.reporter.slf4j.interval\:\ 5\ MINUTE
+kubernetes.operator.reconcile.interval:\ 15\ s
+kubernetes.operator.observer.progress-check.interval:\ 5\ s
+kubernetes.operator.leader-election.enabled:\ true
+kubernetes.operator.leader-election.lease-name:\ flink-operator-lease" \
+-n flink
+```
+Note: The `replicas` config is to set the number of operator replicas.
```
Comment (tomncooper) [inline ha-example/README.md:32]:
```suggestion
Note: The `replicas` configuration is to set to the total number of Flink Kubernetes operator replicas you want.
This will include the leader plus the number of standby instances you want. 
One standby (two replicas in total) is usually sufficient.
```

File: ha-example/README.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+After running this command, you should see there are 2 operator replicas lanuched.
```
Comment (tomncooper) [inline ha-example/README.md:34]:
```suggestion
After running this command, you should see that there are two instances of the Flink Kubernetes operator running.
```

File: ha-example/README.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+```
+
+And when checking the `lease` resource, we can see hte holder(leader) operator's name.
+```
+kubectl get lease -n flink
+NAME                       HOLDER                                      AGE
+flink-operator-lease       flink-kubernetes-operator-6cd86cc8-g298v    45h
+```
+
+When trying to delete the leader operator pod, we can see the holder will change to the other one (or the newly created one),
```
Comment (tomncooper) [inline ha-example/README.md:49]:
```suggestion
You can test the high availability fail-over between the instances, by deleting the leader operator pod.
You should then see that the holder will change to the other instance (or the newly created instance if the creation happens quickly).
```

File: ha-example/README.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+which makes the operator highly available.
+
+## job manager
+
+The JobManager ensures consistency during recovery across TaskManagers. For the JobManager itself to recover consistently,
```
[the remark this answers is not in the document]
Comment (tomncooper) [reply ha-example/README.md:54]: If you make it one sentence per line then it is easier to comment. Also I suggest using soft not hard wrap in your editor, to prevent lines getting split.

File: ha-example/minio-install/README.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+   kubectl port-forward pod/minio 9000 9090 -n flink
+   ```
+3. Create a bucket via MinIO S3 Console
+
+   Open `http://localhost:9090`, and login with `minioadmin/minioadmin`, then go to `Buckets` -> `Create Bucket` to create a bucket.
+4. Monitor the files in the bucket
+
+   Click on the `Object Browser` to view the files in the buckets.
+
+After minio is deployed and bucket is created, the flink configuration can be set like this:
```
Comment (tomncooper) [inline ha-example/minio-install/README.md:20]: How would a user set this, via Helm?

**Review by tomncooper — CHANGES_REQUESTED** [review]
Thanks for looking at this @showuon.

I left some style/grammer comments. Other than that the only thing I think needs adding is examples (or pointers to similar) of how to set the various configs you talk about.

File: ha-example/README.md (hunk 1 above)
Comment (showuon) [reply ha-example/README.md:15]: Yes, it is. Let me update it. Thanks.

File: ha-example/README.md (hunk 2 above)
Comment (showuon) [reply ha-example/README.md:17]: +1! Will work on that!

File: ha-example/README.md [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+   ```
+4. There will be 2 recommendation-app (job manager) and 1 task manager pods deployed
+   ```
+   kubectl  get pod -l app=recommendation-app -n flink
+   NAME                                  READY   STATUS    RESTARTS   AGE
+   recommendation-app-76b6854f98-4qcz4   1/1     Running   0          3m59s
+   recommendation-app-76b6854f98-9zb24   1/1     Running   0          3m59s
+   recommendation-app-taskmanager-1-1    1/1     Running   0          2m5s
+   ``` 
+5. Browse the minio console, to make sure the metadata of the job manager is uploaded to s3://test/ha
```
[the remark this answers is not in the document]
Comment (showuon) [reply ha-example/README.md:97]: Yes, I've documented in `minio-install/README.md`.

File: ha-example/README.md [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+When trying to delete the leader operator pod, we can see the holder will change to the other one (or the newly created one),
+which makes the operator highly available.
+
+## job manager
+
+The JobManager ensures consistency during recovery across TaskManagers. For the JobManager itself to recover consistently,
+an external service must store a minimal amount of recovery metadata (like “ID of last committed checkpoint”),
+as well as help to elect and lock which JobManager is the leader (to avoid split-brain situations).
+
+Configure job manager high availability with kubernetes mode like this in flink configuration:
```
[the remark this answers is not in the document]
Comment (showuon) [reply ha-example/README.md]: Updated to this: 

In order to configure Job Managers in your Flink Cluster for high availability you need to add the following settings to the configuration in your `FlinkDeplyment` CR like this:
```yaml
apiVersion: flink.apache.org/v1beta1
kind: FlinkDeployment
metadata:
  name: recommendation-app
spec:
  image: quay.io/streamshub/flink-sql-runner:v0.0.1
  flinkVersion: v1_19
  flinkConfiguration:
    # job manager HA settings
    high-availability.type: KUBERNETES
    high-availability.storageDir: s3://test/ha
```

File: ha-example/README.md [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+high-availability.type: kubernetes
+high-availability.storageDir: s3:///flink/recovery
+```
+
+## task manager
+
+Checkpointing is Flink’s primary fault-tolerance mechanism, wherein a snapshot of your job’s state persisted periodically to some durable location.
+In the case of failure, Flink will restart from the most recent checkpoint and resume processing.
+Although this is not completely related to HA of the flink cluster, it is also important to know it for the Flink fault tolerance.
+By default, the checkpointing is not enabled in Flink cluster, you can enable it by setting the checkpointing interval:
```
[the remark this answers is not in the document]
Comment (showuon) [reply ha-example/README.md]: Updated to this:

By default, the checkpointing is not enabled in Flink cluster, you can enable it by setting the checkpointing interval in the `FlinkDeplyment` CR like this:
```yaml
apiVersion: flink.apache.org/v1beta1
kind: FlinkDeployment
metadata:
  name: recommendation-app
spec:
  image: quay.io/streamshub/flink-sql-runner:v0.0.1
  flinkVersion: v1_19
  flinkConfiguration:
    # job manager HA settings
    execution.checkpointing.interval: 1min
    state.checkpoints.dir: s3://test/cp
```

File: ha-example/README.md [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+```
+execution.checkpointing.interval: 1min
+state.checkpoints.dir: s3://test/cp
+```
+In this example, it will checkpoint the process status every 1 minute into the s3 path.
+
+### Run HA in recommendation-app application
+
+Here, we will use [recommendation-app](../recommendation-app) as an example to demonstrate the job manager HA.
```
[the remark this answers is not in the document]
Comment (showuon) [reply ha-example/README.md]: I've added step 1 to this:

1. Installing Flink Kubernetes Operator with leader election enabled like this:
   ```
   helm install flink-kubernetes-operator flink-operator-repo/flink-kubernetes-operator \
   --set podSecurityContext=null \
   --set defaultConfiguration."log4j-operator\.properties"=monitorInterval\=30 \
   --set defaultConfiguration."log4j-console\.properties"=monitorInterval\=30 \
   --set replicas=2 \
   --set defaultConfiguration."flink-conf\.yaml"="kubernetes.operator.metrics.reporter.slf4j.factory.class\:\ org.apache.flink.metrics.slf4j.Slf4jReporterFactory
   kubernetes.operator.metrics.reporter.slf4j.interval\:\ 5\ MINUTE
   kubernetes.operator.reconcile.interval:\ 15\ s
   kubernetes.operator.observer.progress-check.interval:\ 5\ s
   kubernetes.operator.leader-election.enabled:\ true
   kubernetes.operator.leader-election.lease-name:\ flink-operator-lease" \
   -n flink
   ```

File: ha-example/minio-install/README.md (hunk 7 above)
Comment (showuon) [reply ha-example/minio-install/README.md]: Updated to this:

After minio is deployed and bucket is created, the flink configuration can be set like this in the `FlinkDeployment` CR:
```yaml
apiVersion: flink.apache.org/v1beta1
kind: FlinkDeployment
metadata:
  name: recommendation-app
spec:
  image: quay.io/streamshub/flink-sql-runner:v0.0.1
  flinkVersion: v1_19
  flinkConfiguration:
    # minio setting 
    s3.access-key: minioadmin
    s3.secret-key: minioadmin
    s3.endpoint: http://MINIO_POD_ID:9000
    s3.path.style.access: "true"
    high-availability.storageDir: s3://test/ha
    state.checkpoints.dir: s3://test/cp
```

File: ha-example/minio-install/minio.yaml [hunk 12]
```diff
@@ -0,0 +1,23 @@
+apiVersion: v1
+kind: Pod
```
[the remark this answers is not in the document]
Comment (showuon) [reply ha-example/minio-install/minio.yaml]: Sure. Updated to using `Deployment`.

**Comment by showuon** [comment]
@tomncooper , thanks for reviewing the PR. I've updated the PR to address your comments. Thanks.

File: ha-example/README.md [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+You should then see that the holder will change to the other instance (or the newly created instance if the creation happens quickly).
+
+## Flink Job Manager
+
+The Job Manager ensures consistency during recovery across Task Managers.
+For the Job Manager itself to recover consistently,
+an external service must store a minimal amount of recovery metadata (like “ID of last committed checkpoint”),
+as well as information needed to elect and lock which Job Manager is the leader (to avoid split-brain situations).
+
+In order to configure Job Managers in your Flink Cluster for high availability you need to add the following settings to the configuration in your `FlinkDeplyment` CR like this:
```
Comment (tomncooper) [inline ha-example/README.md:63]:
```suggestion
In order to configure Job Managers in your Flink Cluster for high availability you need to add the following settings to the configuration in your `FlinkDeployment` CR like this:
```

File: ha-example/README.md [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+    # job manager HA settings
+    high-availability.type: KUBERNETES
+    high-availability.storageDir: s3://test/ha
+```
+
+## Task Manager
+
+[Checkpointing](https://nightlies.apache.org/flink/flink-docs-release-1.20/docs/dev/datastream/fault-tolerance/checkpointing/) is Flink’s primary fault-tolerance mechanism, wherein a snapshot of your job’s state is persisted periodically to some durable location.
+In the case of failure, of a Task running your job's code, Flink will restart the Task from the most recent checkpoint and resume processing.
+Although not strictly related to HA of the Flink cluster, it is important to enable check-pointing in production deployments to ensure fault tolerance.
```
Comment (tomncooper) [inline ha-example/README.md:82]:
```suggestion
Although not strictly related to HA of the Flink cluster, it is important to enable checkpointing in production deployments to ensure fault tolerance.
```
Looks like we are using `checkpointing` in most places rather than `check-pointing`.

File: ha-example/README.md [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+By default, the checkpointing is not enabled in Flink cluster, you can enable it by setting the checkpointing interval in the `FlinkDeplyment` CR like this:
```
Comment (tomncooper) [inline ha-example/README.md:83]:
```suggestion
By default, the checkpointing is not enabled.
You can enable it by setting the checkpointing interval in the `FlinkDeployment` CR like this:
```

File: ha-example/README.md [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+   ```
+5. There should be 2 recommendation-app job manager and 1 task manager pod deployed
+   ```
+   kubectl  get pod -l app=recommendation-app -n flink
+   NAME                                  READY   STATUS    RESTARTS   AGE
+   recommendation-app-76b6854f98-4qcz4   1/1     Running   0          3m59s
+   recommendation-app-76b6854f98-9zb24   1/1     Running   0          3m59s
+   recommendation-app-taskmanager-1-1    1/1     Running   0          2m5s
+   ``` 
+6. Browse the minio console, to make sure the metadata of the job manager is uploaded to s3://test/ha
```
Comment (tomncooper) [inline ha-example/README.md:135]:
```suggestion
6. Browse the minio console, to make sure the metadata of the job manager is uploaded to `s3://test/ha` 
```

File: ha-example/README.md [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+   kubectl  get pod -l app=recommendation-app -n flink
+   NAME                                  READY   STATUS    RESTARTS   AGE
+   recommendation-app-76b6854f98-4qcz4   1/1     Running   0          3m59s
+   recommendation-app-76b6854f98-9zb24   1/1     Running   0          3m59s
+   recommendation-app-taskmanager-1-1    1/1     Running   0          2m5s
+   ``` 
+6. Browse the minio console, to make sure the metadata of the job manager is uploaded to s3://test/ha 
+7. Find out which job manager pod is the leader
+
+   Check the pod logs, if it's the standby job manager, the log will stay at these lines, and waiting for becoming the leader.
```
Comment (tomncooper) [inline ha-example/README.md:138]:
```suggestion
   To do this, check the pod logs. 
   If a job manager is the standby, then the log will indicate that it is watching and waiting to becoming the leader.
```

File: ha-example/README.md [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+   ```
+   kubectl  logs -n flink recommendation-app-76b6854f98-4qcz4 --tail=2 -f
+   ...
+   2024-12-11 08:50:16,438 INFO  org.apache.flink.kubernetes.kubeclient.resources.KubernetesLeaderElector [] - New leader elected  for recommendation-app-cluster-config-map.
+   2024-12-11 08:50:16,518 INFO  org.apache.flink.kubernetes.kubeclient.resources.KubernetesLeaderElector [] - New leader elected 907d6eda-7619-45c5-97c6-e2a6243f56f6 for recommendation-app-cluster-config-map.
+   2024-12-11 08:50:16,524 INFO  org.apache.flink.runtime.jobmaster.MiniDispatcherRestEndpoint [] - http://10.244.1.30:8081 was granted leadership with leaderSessionID=87a6a102-c77b-4ec2-b263-b20a60921c9e
+   2024-12-11 08:50:16,524 INFO  org.apache.flink.runtime.resourcemanager.ResourceManagerServiceImpl [] - Resource manager service is granted leadership with session id 87a6a102-c77b-4ec2-b263-b20a60921c9e.
+   ```
+   You should see the leadership is changed to the other pod.
+9. Make sure the checkpointing file is successfully uploaded onto s3://test/cp via minio console 9Monitor the sink topic in kafka
```
Comment (tomncooper) [inline ha-example/README.md:156]:
```suggestion
9. Make sure the checkpointing file is successfully uploaded to `s3://test/cp` via the minio console.
    Monitor the sink topic in kafka
```

**Review by tomncooper — APPROVED** [review]
LGTM, just a couple of nits.

File: ha-example/README.md [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+```
+helm install flink-kubernetes-operator flink-operator-repo/flink-kubernetes-operator \
+--set podSecurityContext=null \
+--set defaultConfiguration."log4j-operator\.properties"=monitorInterval\=30 \
+--set defaultConfiguration."log4j-console\.properties"=monitorInterval\=30 \
+--set replicas=2 \
+--set defaultConfiguration."flink-conf\.yaml"="kubernetes.operator.metrics.reporter.slf4j.factory.class\:\ org.apache.flink.metrics.slf4j.Slf4jReporterFactory
+kubernetes.operator.metrics.reporter.slf4j.interval\:\ 5\ MINUTE
+kubernetes.operator.reconcile.interval:\ 15\ s
+kubernetes.operator.observer.progress-check.interval:\ 5\ s
```
Comment (Frawless) [inline ha-example/README.md:29]: Are these configurations somehow related to HA configuration (except replicas)? It doesn't look like yes to me and it might be confusing for anyone.

File: ha-example/README.md [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+
+In order to configure Job Managers in your Flink Cluster for high availability you need to add the following settings to the configuration in your `FlinkDeplyment` CR like this:
+```yaml
+apiVersion: flink.apache.org/v1beta1
+kind: FlinkDeployment
+metadata:
+  name: recommendation-app
+spec:
+  image: quay.io/streamshub/flink-sql-runner:v0.0.1
+  flinkVersion: v1_19
```
Comment (Frawless) [inline ha-example/README.md:71]: We should unify flink version here and in docs bellow (1.19 vs 1.20)

File: ha-example/README.md [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+   ```
+   helm install flink-kubernetes-operator flink-operator-repo/flink-kubernetes-operator \
+   --set podSecurityContext=null \
+   --set defaultConfiguration."log4j-operator\.properties"=monitorInterval\=30 \
+   --set defaultConfiguration."log4j-console\.properties"=monitorInterval\=30 \
+   --set replicas=2 \
+   --set defaultConfiguration."flink-conf\.yaml"="kubernetes.operator.metrics.reporter.slf4j.factory.class\:\ org.apache.flink.metrics.slf4j.Slf4jReporterFactory
+   kubernetes.operator.metrics.reporter.slf4j.interval\:\ 5\ MINUTE
+   kubernetes.operator.reconcile.interval:\ 15\ s
+   kubernetes.operator.observer.progress-check.interval:\ 5\ s
```
Comment (Frawless) [inline ha-example/README.md:113]: Same as above

File: ha-example/minio-install/README.md [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+   kubectl apply -f minio.yaml -n flink
+   ```
+2. Access the MinIO S3 Console
+   ```
+   kubectl port-forward deployment/minio 9000 9090 -n flink
+   ```
+3. Create a bucket via MinIO S3 Console
+
+   Open `http://localhost:9090`, and login with `minioadmin/minioadmin`, then go to `Buckets` -> `Create Bucket` to create a bucket.
+4. Monitor the files in the bucket
```
Comment (Frawless) [inline ha-example/minio-install/README.md:16]: Wouldn't be better to provide cli commands for that? There is `mc` command in minio container. This is just an idea, I have no strong opinion.

File: ha-example/README.md [hunk 23]
```diff
@@ -0,0 +1,198 @@
...
+
+After running this command, you should see that there are two instances of the Flink Kubernetes operator running.
+```
+kubectl  get pod -n flink | grep flink-kubernetes-operator
+NAME                                          READY   STATUS    RESTARTS         AGE
+flink-kubernetes-operator-6cd86cc8-fmb2x      2/2     Running   0                3h
+flink-kubernetes-operator-6cd86cc8-g298v      2/2     Running   0                3h
+```
+
+And when checking the `lease` resource, we can see hte holder(leader) operator's name.
```
Comment (tinaselenge) [inline ha-example/README.md:46]:
```suggestion
And when checking the `lease` resource, we can see the holder(leader) operator's name.
```

File: ha-example/README.md (hunk 15 above)
Comment (tinaselenge) [inline ha-example/README.md:83]:
```suggestion
By default, the checkpointing is not enabled in Flink cluster, you can enable it by setting the checkpointing interval in the `FlinkDeployment` CR like this:
```

File: ha-example/README.md (hunk 14 above)
Comment (tinaselenge) [inline ha-example/README.md:82]:
```suggestion
Although not strictly related to HA of the Flink cluster, it is important to enable checkpointing in production deployments to ensure fault tolerance.
```

File: ha-example/README.md [hunk 24]
```diff
@@ -0,0 +1,198 @@
...
+
+   Check the pod logs, if it's the standby job manager, the log will stay at these lines, and waiting for becoming the leader.
+   ```
+   kubectl  logs -n flink recommendation-app-76b6854f98-4qcz4 --tail=2 -f
+   2024-12-11 08:31:22,729 INFO  org.apache.flink.kubernetes.kubeclient.resources.KubernetesConfigMapSharedInformer [] - Starting to watch for flink/recommendation-app-cluster-config-map, watching id:3c7e23f2-fcaa-4f47-8623-fa1ebe9609ea
+   2024-12-11 08:31:22,729 INFO  org.apache.flink.kubernetes.kubeclient.resources.KubernetesConfigMapSharedInformer [] - Starting to watch for flink/recommendation-app-cluster-config-map, watching id:ea191b50-a1ce-43c6-9bfc-a61f739fa8b3
+   ```
+8. Delete the leader pod of job manager, and monitor the logs of the standby pod
+
+   Keep the step(5) command running, and delete the leader pod in another terminal
```
Comment (tinaselenge) [inline ha-example/README.md:146]: do you mean step(7) here?

File: ha-example/README.md (hunk 18 above)
Comment (tinaselenge) [inline ha-example/README.md:156]: 9Monitor?

File: ha-example/README.md (hunk 13 above)
Comment (showuon) [reply ha-example/README.md:63]: Oh, nice catch!

File: ha-example/README.md (hunk 23 above)
Comment (showuon) [reply ha-example/README.md:46]: Oh, nice catch. Thanks.

File: ha-example/README.md (hunk 24 above)
Comment (showuon) [reply ha-example/README.md:146]: Nice catch!

File: ha-example/README.md (hunk 20 above)
Comment (showuon) [reply ha-example/README.md:71]: Good point! Updated.

File: ha-example/minio-install/README.md (hunk 22 above)
Comment (showuon) [reply ha-example/minio-install/README.md:16]: Because we are running the examples manually, I think it'd be easier to check via UI, not CLI.

**Comment by showuon** [comment]
PR updated. Thanks for the comments!

**Comment by showuon** [comment]
If there are no more comments, I'm going to merge it today. Thanks.

**Review by Frawless — APPROVED** [review]
